File: lib/data/source/sync/sync_transport.dart
/// The byte-transport seam between the sync protocol (pure Dart) and the actual
/// carrier (Bluetooth RFCOMM via `bluetooth_sync_native`).
///
/// The protocol layer ([SyncSession]) depends ONLY on [SyncByteTransport], so it
/// can be driven over an in-memory pipe in tests with no Bluetooth at all. The
/// production implementation lives in the data layer and wraps the plugin's
/// `sendBytes` / `onBytesReceived`.
library;

import 'dart:async';
import 'dart:typed_data';

/// A bidirectional, ordered, reliable byte channel to one peer. Bytes written
/// with [send] arrive at the peer's [inbound] stream; the carrier may re-chunk
/// them arbitrarily (framing handles that). [inbound] closes when the link ends.
abstract interface class SyncByteTransport {
  /// Sends [bytes] to the peer. Completes when the bytes have been handed to the
  /// carrier (which applies backpressure over a slow link).
  Future<void> send(Uint8List bytes);

  /// Raw inbound byte chunks from the peer, in order. Closes on disconnect.
  Stream<Uint8List> get inbound;

  /// Closes the channel. Idempotent.
  Future<void> close();
}

/// An in-memory [SyncByteTransport] pair for tests: two endpoints wired together
/// so what one [send]s appears on the other's [inbound]. Models the real carrier
/// closely enough to exercise the whole protocol without Bluetooth.
///
/// Use [SyncPipe.create] to get the connected `(a, b)` pair.
class SyncPipe implements SyncByteTransport {
  SyncPipe._(this._name);

  /// Creates two endpoints wired to each other.
  static (SyncPipe, SyncPipe) create() {
    final a = SyncPipe._('A');
    final b = SyncPipe._('B');
    a._peer = b;
    b._peer = a;
    return (a, b);
  }

  final String _name;
  late final SyncPipe _peer;
  final StreamController<Uint8List> _inbound =
      StreamController<Uint8List>.broadcast();
  bool _closed = false;

  @override
  Stream<Uint8List> get inbound => _inbound.stream;

  @override
  Future<void> send(Uint8List bytes) async {
    if (_closed) throw StateError('pipe $_name is closed');
    // Deliver on a microtask so send/receive interleave like a real async link
    // rather than re-entering the peer synchronously.
    final copy = Uint8List.fromList(bytes);
    scheduleMicrotask(() {
      if (!_peer._closed && !_peer._inbound.isClosed) {
        _peer._inbound.add(copy);
      }
    });
  }

  @override
  Future<void> close() async {
    if (_closed) return;
    _closed = true;
    if (!_inbound.isClosed) await _inbound.close();
    // Closing one end ends the peer's inbound too.
    if (!_peer._closed && !_peer._inbound.isClosed) {
      await _peer._inbound.close();
      _peer._closed = true;
    }
  }
}

File: app/src/main/kotlin/tech/mmarca/openvitals/features/devicesync/protocol/SyncTransport.kt
package tech.mmarca.openvitals.features.devicesync.protocol

import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.channels.ReceiveChannel

/**
 * The byte-transport seam between the protocol and the carrier. The
 * protocol depends only on this, so tests drive it over an in-memory pipe.
 */
interface SyncByteTransport {
    /** Sends [bytes]. Suspends until handed to the carrier. Concurrent sends are not safe. */
    suspend fun send(bytes: ByteArray)

    /** Inbound chunks in order, re-chunked arbitrarily. Closes when the link ends. */
    val inbound: ReceiveChannel<ByteArray>

    /** Closes the channel. Idempotent. */
    fun close()
}

/** An in-memory transport pair for tests. Use [SyncPipe.create]. */
class SyncPipe private constructor() : SyncByteTransport {

    private lateinit var peer: SyncPipe
    private val inboundChannel = Channel<ByteArray>(Channel.UNLIMITED)

    @Volatile private var closed = false

    override val inbound: ReceiveChannel<ByteArray> get() = inboundChannel

    override suspend fun send(bytes: ByteArray) {
        check(!closed) { "pipe is closed" }
        // Buffered channel: delivery never re-enters the peer synchronously.
        peer.inboundChannel.trySend(bytes.copyOf())
    }

    override fun close() {
        if (closed) return
        closed = true
        inboundChannel.close()
        // Closing one end ends the peer's inbound too.
        if (!peer.closed) {
            peer.closed = true
            peer.inboundChannel.close()
        }
    }

    companion object {
        /** Creates two endpoints wired to each other. */
        fun create(): Pair<SyncPipe, SyncPipe> {
            val a = SyncPipe()
            val b = SyncPipe()
            a.peer = b
            b.peer = a
            return a to b
        }
    }
}
